rfc/0000-lock-file.md

```diff
@@ -0,0 +1,280 @@
+
+```toml
+# Marker.toml
+
+[lints]
+lint-crate = { path = "./lint-crate" }
+marker_lints = "0.4.0"
+
+[lints.marker_lints.config]
+config_foo = "bar"
+```
+
+I like `Marker.toml` because it also gives `Marker` more visibility. When looking at the repo files this gives you an instant recognition that this project uses Marker. It also makes `Marker.toml` and `Marker.lock` paired just like `Cargo.toml` and `Cargo.lock`.
+
+**Example:**
+
+Suppose you have your own `lint-crate` inside of the workspace that needs to be linted and a dependency on `marker_lints` crate from `crates.io`. The folder structure of such a workspace would look like the following.
+
+- `workspace-crate/`: directory with a regular Rust crate that you develop as part of your app
+- `lint-crate/`
+    - `Cargo.toml`: declares a dependency on `marker_api`
+    - `src/`
+        - `lib.rs`: defines lints, configs, implements the lint passes
+- `Cargo.toml`: declares `workspace-crate` and `lint-crate` as `workspace.members`
+- `Cargo.lock`: contains entries for `lint-crate` and its `marker_api` dependency
+- `Marker.toml`: defines `lint-crate` and `marker_lints` under the `[lints]` dependencies table
+- `Marker.lock`: hardlink to `target/marker/Cargo.lock` and contains only `marker_lints` and its `marker_api` entries
+- `target/`
+    - `marker/`
+        - `Cargo.toml`: contains a definition of the `buildspace` and the "dispatch" crate. The `[dependencies]` section includes only `marker_lints = "x.y.z"`. The `path` dependencies never get into the `dependencies` here.
+
+### `Marker.lock`
+
+The `Marker.lock` file is a hardlink to the internal buildspace's `Cargo.lock` file. The internal buildspace structure is not stable yet, and not exposed fully to the users, but you may expect it to have a `Cargo.lock` file, and apply any commands that manage the lock file or read the metadata.
+
+The `cargo marker` CLI exposes a command `cargo marker buildspace` that can be suffixed with any `cargo` command and it will be run inside of the internal Marker buildspace to manage the lock file, licensing, do crates pre-fetching for caching etc. For example, you may run commands like the following.
+
+- `cargo marker buildspace update --package foo`
+- `cargo marker buildspace deny`
+- `cargo marker buildspace tree --depth 1`
+- `cargo marker buildspace generate-lockfile`
+- `cargo marker buildspace fetch`
+
+The commands that may be run inside of the buildspace this way aren't limited to make possible the usage of external cargo plugins that manage the lockfile. We'll however try to maintain a list of commands that are known to be safe to use in the docs.
+
+# Reference-level explanation
+
+## Buildspace
+
+The `cargo marker` CLI will always try to create a hardlink to `Marker.lock` in the internal buildspace using `cargo generate-lockfile` if it doesn't exist and there are non-`path` dependencies. The users will decide on their own if they want to commit this file to their source control or not.
```

Comment:
I've never had such a use case, and `cargo` doesn't have such a parameter. I'm sure it's possible to have such a parameter implemented, but I'd keep things simple, and follow cargo in this regard. If there arises a compelling use case for this, we'll revisit this

Reply:
I was mainly thinking of the CLI use case when I suggested this. I only later noticed, that this was the one I was considering. Otherwise, there is little use for such an option